Move json encoding out of riak_object

engelsanchez opened the pull request:
Revisiting https://github.com/basho/riak_kv/pull/88
But using riak_object_json encode/decode and leaving old methods in
riak_object.erl with a deprecation warning.

engelsanchez committed:
Move json encoding out of riak_object

Revisiting https://github.com/basho/riak_kv/pull/88
But using riak_object_json encode/decode and leaving old methods in
riak_object.erl with a deprecation warning.

rzezeski commented:
- [x] verify compile
- [x] verify unit test/EQC
- [x] verify copy pasta
- [x] verify specs
- [x] while I'm here, check for any dumb code ( that has survived, not that you introduced Engel ;) )
- [x] verify riak test
- [ ] merge
- [x] close #88

rzezeski commented on src/riak_object_json.erl:
```diff
@@ -0,0 +1,183 @@
+
+-export([encode/1,decode/1]).
+
+%% @spec to_json(riak_object()) -> {struct, list(any())}
```
Kill this doc spec.

rzezeski commented on src/riak_object_json.erl:
```diff
@@ -0,0 +1,183 @@
+                 end,
+    case lists:foldl(Classifier, undefined, List) of
+        struct -> {struct, jsonify_proplist(List)};
+        string -> list_to_binary(List);
```
I need to look at this more closely but right off the bat wonder if we have potential utf-8 issues here?

Granted, it is the same as current behavior.  But making mental note to check.

engelsanchez replied:
I don't see how we could have issues in this layer: UTF-8 or ASCII strings should look identical to this code (integer values between 0-255) and any errors would be in other layers interpreting this data. The funky thing I noticed now is that we are allowing 256!!! This code will probably never see one of those, but I think the correct thing is < 256 or =< 255

rzezeski commented on src/riak_object_json.erl:
```diff
@@ -0,0 +1,183 @@
+jsonify_proplist([]) -> [];
+jsonify_proplist(List) ->
+    dict:to_list(lists:foldl(fun({Key, Value}, Dict) ->
+                                     JSONKey = if is_list(Key) -> list_to_binary(Key);
```
Once again, wonder about potential utf-8 issues.

rzezeski commented on src/riak_object_json.erl:
```diff
@@ -0,0 +1,183 @@
+                  {<<"data">>, D}|T], Accum) ->
+    Converter = fun({Key, Val}) ->
+                        case Key of
+                            <<"Links">> ->
```
`<<"Links">>` and `<<"X-Riak-Last-Modified">>` could be replaced with their macros in `riak_kv_wm_raw.hrl`.

engelsanchez committed:
Addressing review comments

Using constants for links and last modified headers.
Removing old @spec.
String detection should not allow 256 int values.

engelsanchez commented:
@rzezeski killed the spec, used the header constants and tweaked the string detection code. I don't _think_ I see potential UTF-8 issues, but let me know if you think there's something there.

seancribbs commented:
Couldn't those type guard BIFs be moved out of guards and replaced with unicode:c2l or c2b? Then you could at least verify that they are valid UTF-8.

Sean Cribbs

On Jul 12, 2013, at 8:59 AM, Engel Sanchez notifications@github.com wrote:

> @rzezeski killed the spec, used the header constants and tweaked the string detection code. I don't think I see potential UTF-8 issues, but let me know if you think there's something there.
> 
> —
> Reply to this email directly or view it on GitHub.

engelsanchez commented:
@rzezeski @seancribbs I'm not opposed to improve UTF-8 problem detection here, but could we open another issue for that? It's only taken 2 years to merge this simple refactor :)

seancribbs commented:
@engelsanchez Yeah, you're right. I think longer-term the better way would be to constrain the names/types of metadata to avoid the problem, but that's too much for this PR.

rzezeski commented:
+1 to merge
